src/include/plan_node/abstract_join_plan_node.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+#include "parser/expression/abstract_expression.h"
+#include "plan_node/abstract_plan_node.h"
+
+namespace terrier::plan_node {
+
+//===--------------------------------------------------------------------===//
+// Abstract Join Plan Node
+//===--------------------------------------------------------------------===//
+
+class AbstractJoinPlanNode : public AbstractPlanNode {
+ public:
+  AbstractJoinPlanNode(std::shared_ptr<OutputSchema> output_schema, LogicalJoinType join_type,
+                       parser::AbstractExpression *predicate)
+      : AbstractPlanNode(std::move(output_schema)), join_type_(join_type), predicate_(predicate) {}
+
+  common::hash_t Hash() const override;
+
+  bool operator==(const AbstractPlanNode &rhs) const override;
+  bool operator!=(const AbstractPlanNode &rhs) const override { return !(*this == rhs); }
+
+  //===--------------------------------------------------------------------===//
+  // Accessors
+  //===--------------------------------------------------------------------===//
+
+  LogicalJoinType GetLogicalJoinType() const { return join_type_; }
+
+  const parser::AbstractExpression *GetPredicate() const { return predicate_; }
+
+ private:
+  LogicalJoinType join_type_;
+
+  const parser::AbstractExpression *predicate_;
+
+  DISALLOW_COPY_AND_MOVE(AbstractJoinPlanNode);
+};
+
```

Comment:
The general c++ approach is to write a functor to do hashing instead of a virtual method. It probably doesn't matter that much for plan nodes since they need a vtable anyway, but it might be beneficial to write a `std::hash` specialization so you can use these classes easily with STL. Find example in https://github.com/cmu-db/terrier/blob/master/src/include/common/strong_typedef.h#L333

Reply:
Agreed. I think it would be important to start doing this more.